### PR TITLE
PROJECT: ASSOCIATIONS

### PR DESCRIPTION
**Pull request by bmuhamood**
In this project, we created three tables (User, Event, and attendance), and they are connected using the has_many_through relationships on the models.
We build a site in which a user can create an account with that the user can create as many events as he/she wishes, at the same time the user can attend different events created by other users, the attendances are store on their table, we also used sessions to create our own login and logout actions.
While 'logged-in' the user can: create events, see all events, and attend other events.
While 'logged-out' the user can: see all events, create an account, or if he/she has an account already log-in.

### CODE REVIEW

**Review by trekab — CHANGES_REQUESTED** [review]
### Changes Required! :recycle:
Hi team! Well done so far. There are some vital requirements that I'd like you to consider before moving forward though. These include:
- [x] Add more descriptive instructions about setting up your project locally. The usual `bundle install`, `rails db:create`, and `rails db:migrate` and really helpful when it comes to instructing the user trying out your app locally.
- [x] __[Optional]__ Much as you have added a pull request summary, it appears that you directly copied and pasted text from the odin project without personalizing it. This makes the tense of your message wrong. Try to personalize this and think of the PR description as a summary of changes introduced by your commits to the project.
- [x] Navigation links should be present at the top of the page ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/navigation_links_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations)). For most of your routes, I have to manually type them in my browser in order to access the resources. The navigation links should be able to guide the user.
- [x] The EventsController's `create` action should use the `build` method on association, i.e. `current_user.events.build(event_params)` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/build_method_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
- [x] The UsersController's `show` action should:
  - list events created by the user, i.e. `@created_events = current_user.events` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/current_user_show_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
  - list events previoulsy attendented by user **(Hint: use event model scope)**, i.e. `@past_events = current_user.attended_events.previous` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
  - list upcoming events previously attended by user **(Hint: use event model scope)**, i.e. `@upcoming_events = current_user.attended_events.upcoming` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
- [x] Add indexes for your database foreign keys ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/foreign_key_index_requirement.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49723079/)). This is missing for the `creator_id` field on the `events` table.
- [x] Please add tests for your project using Rspec. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/unit_tests_requirement.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49723079/))

__Kindly make the necessary changes and submit another code review request.__
_Feel free to reach me on slack for any inquiries about this review._ :wink: 
__Treasure Kabareebe__
[GitHub](https://github.com/trekab) | [LinkedIn](https://www.linkedin.com/in/treasure-kabareebe/) | [Twitter](https://twitter.com/TKabareebe)

File: README.md [hunk 1]
```diff
@@ -1 +1,58 @@
...
+
+- Go to the main page of the repo.
+- Press the "Code" button and get the repo link.
+- Clone it using git.
+
+## How To Run This Application In Console
+
+- After cloning the repo, go to the directory where the project was cloned and open your terminal.
+- Once you are on the terminal type `bundle install` 
+- Then run `rails db:migrate` no migrate the database.
```
Comment (maelfosso) [inline README.md]: We can't run `rails db:migrate` without running `rails db:create`
Kindly add it

File: README.md [hunk 2]
```diff
@@ -1 +1,58 @@
...
+## How to install Ruby On Rails
+
+- Follow Ruby on Rails Installation Guide Using this link `https://guides.rubyonrails.org/v5.0/getting_started.html`
+- Once done, go to project directory and test if Rails is working fine on your Terminal by typing ``rails server``
+
+To get a local copy up and running follow these simple example steps.
+
+- Go to the main page of the repo.
+- Press the "Code" button and get the repo link.
+- Clone it using git.
```
Comment (maelfosso) [inline README.md]: Kindly give the command to run

File: app/controllers/events_controller.rb [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+  # GET /events/1/edit
+  def edit; end
+
+  def attending
+    set_event
+    @attendee = Attendance.new
+    @attendee.event_id = @event.id
+    @attendee.user_id = current_user.id
+    @attendee.save
+  end
```
Comment (maelfosso) [inline app/controllers/events_controller.rb]: The fact that there is no view for that action no redirection cause the app to download a kind of file.
Kindly redirect after a successful save. Redirect to the same page for example.

File: app/views/users/show.html.erb [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+  </p>
+
+  <p>
+    <strong>Email:</strong>
+    <%= @user.email %>
+  </p>
+  <%= link_to 'Edit', edit_user_path(@user) %> |
+  <%= link_to 'Home', root_path %>
+</div>
+<h2> This are <%= @user.username %>'s events </h2>
```
Comment (maelfosso) [inline app/views/users/show.html.erb]: It's not **This are** but `**These** are`

File: app/views/events/show.html.erb [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+  <strong>Date:</strong>
+  <%= @event.date %>
+</p>
+<p>
+  <strong>Atendees</strong>
+</p>
+
+<p>
+  <strong>Number of Attendees:</strong>
+  <%= @attendees.where(event_id: @event.id).count %><br><br>
```
Comment (maelfosso) [inline app/views/events/show.html.erb]: Such an expression must be done into a controller, not in the view.
Kindly remove it, please.

File: app/views/events/show.html.erb [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+  <strong>Attendees:</strong>
+  <ol>
+    <% @attendees.where(event_id: @event.id).each do |i|%>
```
Comment (maelfosso) [inline app/views/events/show.html.erb:20]: Such an expression must be done into a controller, not in the view.
Kindly remove it, please.

File: app/models/user.rb [hunk 7]
```diff
@@ -0,0 +1,10 @@
+class User < ApplicationRecord
+  validates :username, :email, presence: true
+  has_many :attendances
+  has_many :events, through: :attendances
+  has_many :own_events, foreign_key: 'creator_id', class_name: 'Event'
+
+  scope :previous, -> { where('date < ?', Date.today) }
+  scope :upcoming, -> { where('date > ?', Date.today) }
+  scope :today, -> { where('date == ?', Date.today) }
```
Comment (maelfosso) [inline app/models/user.rb]: Used nowhere.
Kindly remove it, please.

File: app/views/users/show.html.erb [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+<% @user.own_events.each do |e| %>
```
Comment (maelfosso) [inline app/views/users/show.html.erb]: Create the view variables in the controller and use it here

File: app/views/users/_attended.html.erb [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+      <th>Body</th>
+      <th>Date</th>
+      <th>Creation date</th>
+      <th>Creator</th>
+      <th colspan="3"></th>
+    </tr>
+  </thead>
+
+  <tbody>
+  <% @user.events.previous.each do |event| %>
```
Comment (maelfosso) [inline app/views/users/_attended.html.erb]: Create the view variables in the controller and use it here

File: app/views/users/_attend_events.html.erb [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+      <th>Body</th>
+      <th>Date</th>
+      <th>Creation date</th>
+      <th>Creator</th>
+      <th colspan="3"></th>
+    </tr>
+  </thead>
+
+  <tbody>
+  <% @user.events.upcoming.each do |event| %>
```
Comment (maelfosso) [inline app/views/users/_attend_events.html.erb]: Create the view variables in the controller and use it here

File: app/views/users/_today_events.html.erb [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+      <th>Body</th>
+      <th>Date</th>
+      <th>Creation date</th>
+      <th>Creator</th>
+      <th colspan="3"></th>
+    </tr>
+  </thead>
+
+  <tbody>
+  <% @user.events.today.each do |event| %>
```
Comment (maelfosso) [inline app/views/users/_today_events.html.erb]: Create the view variables in the controller and use it here

**Review by maelfosso — CHANGES_REQUESTED** [review]
# Changes required 🔴 

Hello,

Thank you for the well-done job so far.
However, there are still some issues 

Kindly go through the comments left below, make the suggested changes and submit a new code review request
Thanks 

Mael FOSSO

**Review by javitocor — APPROVED** [review]
# STATUS APPROVED 🥇 🌟 

Congratulations team! 🥳 
You have met all the requirements and your project is approved! 💯 
Remember to fill up the completion form 📑  and you can move into the next project ⏩ 

Good luck 😃

**Pull request merged**